drivers/mipi_dbi/mipi_dbi_nxp_lcdic.c

```diff
@@ -0,0 +1,803 @@
+
+/* Descriptor for LCDIC command */
+union lcdic_trx_cmd {
+	struct {
+		/* Data length in bytes. LCDIC transfers data_len + 1 */
+		uint32_t data_len: 18;
+		/* Dummy SCLK cycles between TX and RX (for SPI mode) */
+		uint32_t dummy_count: 3;
+		uint32_t rsvd: 2;
+		/* Use auto repeat mode */
+		uint32_t auto_repeat: 1;
+		/* Tearing enable sync mode */
+		uint32_t te_sync_mode: 2;
+		/* TRX command timeout mode */
+		uint32_t trx_timeout_mode: 1;
+		/* Data format, see lcdic_data_fmt */
+		uint32_t data_format: 2;
+		/* Enable command done interrupt */
+		uint32_t cmd_done_int: 1;
+		/* LCD command or LCD data, see lcdic_cmd_dc */
+		uint32_t cmd_data: 1;
+		/* TX or RX command, see lcdic_cmd_type */
+		uint32_t trx: 1;
+	} bits;
+	uint32_t u32;
+};
+
+struct mipi_dbi_lcdic_config {
+	LCDIC_Type *base;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	bool swap_bytes;
+};
+
+#ifdef CONFIG_MIPI_DBI_NXP_LCDIC_DMA
+struct stream {
+	const struct device *dma_dev;
+	uint32_t channel;
+	struct dma_config dma_cfg;
+	struct dma_block_config blk_cfg[2];
+};
+#endif
+
+struct mipi_dbi_lcdic_data {
+	/* Tracks number of bytes remaining in command */
+	uint32_t cmd_bytes;
+	/* Tracks number of bytes remaining in transfer */
+	uint32_t xfer_bytes;
+	/* Tracks start of transfer buffer */
+	const uint8_t *xfer_buf;
+	/* When sending data that does not evenly fit into 4 byte chunks,
+	 * this is used to store the last unaligned segment of the data.
+	 */
+	uint32_t unaligned_word __aligned(4);
+	/* Tracks lcdic_data_fmt value we should use for pixel data */
+	uint8_t pixel_fmt;
+	const struct mipi_dbi_config *active_cfg;
+	struct k_sem xfer_sem;
+	struct k_sem lock;
+#ifdef CONFIG_MIPI_DBI_NXP_LCDIC_DMA
+	struct stream dma_stream;
+#endif
+};
+
+#define LCDIC_ALL_INTERRUPTS \
+	(LCDIC_ICR_RFIFO_THRES_INTR_CLR_MASK | \
+	LCDIC_ICR_RFIFO_UNDERFLOW_INTR_CLR_MASK | \
+	LCDIC_ICR_TFIFO_THRES_INTR_CLR_MASK | \
+	LCDIC_ICR_TFIFO_OVERFLOW_INTR_CLR_MASK | \
+	LCDIC_ICR_TE_TO_INTR_CLR_MASK | \
+	LCDIC_ICR_CMD_TO_INTR_CLR_MASK | \
+	LCDIC_ICR_CMD_DONE_INTR_CLR_MASK | \
+	LCDIC_ICR_RST_DONE_INTR_CLR_MASK)
+
+/* RX and TX FIFO thresholds */
+#ifdef CONFIG_MIPI_DBI_NXP_LCDIC_DMA
+#define LCDIC_RX_FIFO_THRESH 0x0
+#define LCDIC_TX_FIFO_THRESH 0x0
+#else
+#define LCDIC_RX_FIFO_THRESH 0x0
+#define LCDIC_TX_FIFO_THRESH 0x3
+#endif
+
+/* Timer0 and Timer1 bases. We choose a longer timer0 base to enable
+ * long reset periods
+ */
+#define LCDIC_TIMER0_RATIO 0xF
+#define LCDIC_TIMER1_RATIO 0x9
+
+/* After LCDIC is enabled or disabled, there should be a wait longer than
+ * 5x the module clock before other registers are read
+ */
+static inline void mipi_dbi_lcdic_reset_delay(void)
+{
+	/* With a core clock of 260MHz, we need a minimum wait of 13
+	 * cycles to satisfy requirement. Wait for 130 as this is
+	 * the same value MCUX SDK uses.
+	 */
```

Comment:
Given this driver can be reused on other SoC's with varying core speeds, is it safe to have a loop like the below. Should we instead use Zephyr delay functions?

Reply:
No- thanks for catching this. I'll switch this to use `k_busy_wait`